### PR TITLE
Create TimeSync Header and Hardware Implementation

### PR DESCRIPTION
**Pull request by Raghav-M-Daga**

### CODE REVIEW

File: Gimbalize/src/hardwareReader.cpp [hunk 1]
```diff
@@ -4,6 +4,10 @@
 #include <Adafruit_BMP085.h>
 #include <SPI.h>
 #include <SD.h>
+#include "timeSync.h"
+
```
Comment (pchou05) [inline Gimbalize/src/hardwareReader.cpp:8]: #include <MKRIMU.h>

File: Gimbalize/src/hardwareReader.cpp [hunk 2]
```diff
@@ -4,6 +4,10 @@
...
+TimeSync bmpTimeSync;
+TimeSync mpuTimeSync;
 
 Adafruit_MPU6050 mpu;
```
Comment (pchou05) [inline Gimbalize/src/hardwareReader.cpp:12]: delete

File: Gimbalize/src/hardwareReader.cpp [hunk 3]
```diff
@@ -84,16 +88,32 @@ void loop() {
   lastTime = currentTime;
 
   sensors_event_t a, g, temp;
+  float reading_time = millis()
```
Comment (namnik100) [inline Gimbalize/src/hardwareReader.cpp:91]: Missing semicolon

File: Gimbalize/src/hardwareReader.cpp [hunk 4]
```diff
@@ -108,21 +128,41 @@ void loop() {
   Serial.print(altitude);
   Serial.println(" m");
 
+  try {
+    desired_time = millis()
```
Comment (namnik100) [inline Gimbalize/src/hardwareReader.cpp:132]: missing semicolon

File: Gimbalize/src/hardwareReader.cpp [hunk 5]
```diff
@@ -108,21 +128,41 @@ void loop() {
...
+    std::array<double, 3> BMP_interpolated_values = bmpTmeSync.getValuesAt(desired_time);
```
Comment (namnik100) [inline Gimbalize/src/hardwareReader.cpp:133]: "bmpTmeSync" undefined

File: Gimbalize/src/hardwareReader.cpp [hunk 6]
```diff
@@ -108,21 +128,41 @@ void loop() {
...
+    std::array<double, 3> MPU_interpolated_values = mpuTmeSync.getValuesAt(desired_time);
```
Comment (namnik100) [inline Gimbalize/src/hardwareReader.cpp:134]: "mpuTmeSync" undefined

File: Gimbalize/src/hardwareReader.cpp (hunk 4 above)
Comment (namnik100) [reply Gimbalize/src/hardwareReader.cpp:132]: "desired_time" never instantiated, made sure to do it outside try and unsigned

File: Gimbalize/src/timeSync.h [hunk 7]
```diff
@@ -0,0 +1,29 @@
...
+
+    // Get interpolated values at the desired time
+    std::array<double, 3> getValuesAt(double desired_time) const;
+
+private:
+    size_t max_history_size;
+    std::vector<std::pair<double, std::array<double, 3>>> readings; // {timestamp, {values}}
+
+    // Helper to find the two closest timestamps to the desired time
+    std::pair<std::pair<double, std::array<double, 3>>, std::pair<double, std::array<double, 3>>> getClosestReadings(double desired_time) const;
```
Comment (namnik100) [inline Gimbalize/src/timeSync.h:26]: getClosestReadings not defined/implemented anywhere?

File: Gimbalize/src/timeSync.h [hunk 8]
```diff
@@ -0,0 +1,29 @@
...
+class TimeSync {
+public:
+    // Constructor: max_history_size controls how many readings are stored
+    TimeSync(size_t max_history_size = 100);
+
+    // Add a new sensor reading (timestamp in ms and values as array)
+    void addReading(double timestamp, const std::array<double, 3>& values);
+
+    // Get interpolated values at the desired time
+    std::array<double, 3> getValuesAt(double desired_time) const;
```
Comment (namnik100) [inline Gimbalize/src/timeSync.h:19]: No edge-case handling for when time is 0 for instance (will return null), cover all edge cases

**Review by namnik100 — CHANGES_REQUESTED** [review]
Check issues for stuff to add as well

File: Gimbalize/src/hardwareReader.cpp (hunk 6 above)
Comment (namnik100) [reply Gimbalize/src/hardwareReader.cpp:134]: You can also store millis() in a variable at beginning of loop and compare with last loop time (mb realized that cell timing wasn't that accurate)

**Comment by namnik100** [comment]
testing webhook ignore

**Review by namnik100 — COMMENTED** [review]
mpuTimeSync() and bmpTimeSynx() not defined
Include timesync.h dependencies in main

the interpolation is assuming that ull get a reading when u ask for it, like the sensor doesnt have to give a value at that time, make sure u throw an exception for this case like if(readings.empty)

delay(100) causes loop overruns use non-blocking logic like with millis()

there's thing i wanted to try call  rolling time window, so instead of using one desired_time specific like a range like 50 miliseconds from desired time, get all of those values, and find the average as the actual value

**Pull request closed without merging**